File: Capítulo_8/Tratando colisões nas tabelas hash/Colisões/hashTable.js
import { defaultToString, ValuePair } from "../../util/util.js";

class HashTable {
    #table;
    constructor(toStrFn = defaultToString) {
        this.toStrFn = toStrFn;
        this.#table = {};
    }

    // Método getter para #table:
    _getTable() {
        return this.#table;
    }

    // Método setter para #table:
    _setTable(key, item) {
        this.#table[key] = item;
    }

    // Método loseloseHashCode:
    loseloseHashCode(key) {
        if (typeof key === "number") {
            return key;
        }
        const tableKey = this.toStrFn(key);
        let hash = 0;
        for (let i = 0; i < tableKey.length; i++) {
            hash += tableKey.charCodeAt(i);
        }
        return hash % 37;
    }

    // Método hashCode:
    hashCode(key) {
        return this.loseloseHashCode(key);
    }

    // put(key,value): esse método adiciona um novo item à tabela hash (ou pode atualizá-la também).
    put(key, value) {
        if (key != null && value != null) {
            const position = this.hashCode(key);
            this._setTable(position, new ValuePair(key, value));
            return true;
        }
        return false;
    }

    //get(key): esse método devolve um value específico encontrado com key.
    get(key) {
        const valuePair = this._getTable()[this.hashCode(key)];
        return valuePair == null ? undefined : valuePair.value;
    }

    // remove(key): esse método remove o value da tabela hash usando key.
    remove(key) {
        const hash = this.hashCode(key);
        const valuePair = this._getTable()[hash];
        if (valuePair != null) {
            delete this._getTable()[hash];
            return true;
        }
        return false;
    }

    size() {
        return Object.keys(this._getTable()).length;
    }

    isEmpty() {
        return this.size() === 0;
    }

    toString() {
        if (this.isEmpty()) {
            return "";
        }
        const keys = Object.keys(this._getTable());
        let objString = `{${keys[0]} => ${this._getTable()[
            keys[0]
        ].toString()}}`;
        for (let i = 1; i < keys.length; i++) {
            objString = `${objString}, {${keys[i]} => ${this._getTable()[
                keys[i]
            ].toString()}}`;
        }
        return objString;
    }
}

// Vamos testar a classe HashTable executando alguns exemplos:
const hash = new HashTable();
hash.put("Ygritte", "ygritte@email.com");
hash.put("Jonathan", "jonathan@email.com");
hash.put("Jamie", "jamie@email.com");
hash.put("Jack", "jack@email.com");
hash.put("Jasmine", "jasmine@email.com");
hash.put("Jake", "jake@email.com");
hash.put("Nathan", "nathan@email.com");
hash.put("Athelstan", "athelstan@email.com");
hash.put("Sue", "sue@email.com");
hash.put("Aethelwulf", "aethelwulf@email.com");
hash.put("Sargeras", "sargeras@email.com");

console.log(hash.size());
console.log(hash.toString());

File: Capítulo_8/Tratando colisões nas tabelas hash/Colisões/Código em Python/hashTable.py
# defaultToString:
def defaul_to_string(item):
    if item is None:
        return "NONE"
    elif isinstance(item, str):
        return f"{item}"
    return str(item)

# Classe ValuePair
class ValuePair:
    def __init__(self, key, value):
        self.key = key
        self.value = value

    def __str__(self):
        return f"[#{self.key}: {self.value}]"
    
# Classe HashTable
class HashTable:
    def __init__(self, to_str_fn = defaul_to_string):
        self.to_str_fn = to_str_fn
        self.__table = {}

    # Método getter para __table:
    def get_table(self):
        return self.__table

    # Método setter para __table:
    def set_table(self, key, item):
        self.__table[key] = item
    
    # Método loselose_hash_code:
    def loselose_hash_code(self,  key):
        if isinstance(key, (int, float)):
            return key
        table_key = self.to_str_fn(key)
        hash = 0
        for i in range(len(table_key)):
            hash += ord(table_key[i])
        return hash % 37
    
    # Método hash_code:
    def hash_code(self, key):
        return self.loselose_hash_code(key)
    
    # put(key,value): esse método adiciona um novo item à tabela hash (ou pode atualizá-la também).
    def put(self, key, value):
        if key and value is not None:
            position = self.hash_code(key)
            self.set_table(position, ValuePair(key, value))
            return True
        return False
    
    # get(key): esse método devolve um value específico encontrado com key.
    def get(self, key):
        hash = self.hash_code(key)
        if hash in self.get_table():
            value_pair = self.get_table()[hash]
            return None if value_pair is None else value_pair.value
        return None
    
    # remove(key): esse método remove o value da tabela hash usando key.
    def remove(self, key):
        hash = self.hash_code(key)
        table = self.get_table()
        if hash in table:
            del table[hash]
            return True
        return False
    
    # Método size:
    def size(self):
        return len(self.get_table())
    
    # Método is_empty:
    def is_empty(self):
        return self.size() == 0
    
    # Método to_string
    def to_string(self):
        if self.is_empty():
            return ""
        keys = list(self.get_table().keys())
        obj_string = f"{{{keys[0]} => {self.get_table()[keys[0]].__str__()}}}"
        for i in range(1, len(keys)):
            obj_string = f"{obj_string}, {{{keys[i]} => {self.get_table()[keys[i]].__str__()}}}"
        return obj_string
    
# Vamos testar a classe HashTable executando alguns exemplos:
hash = HashTable()
hash.put("Ygritte", "ygritte@email.com")
hash.put("Jonathan", "jonathan@email.com")
hash.put("Jamie", "jamie@email.com")
hash.put("Jack", "jack@email.com")
hash.put("Jasmine", "jasmine@email.com")
hash.put("Jake", "jake@email.com")
hash.put("Nathan", "nathan@email.com")
hash.put("Athelstan", "athelstan@email.com")
hash.put("Sue", "sue@email.com")
hash.put("Aethelwulf", "aethelwulf@email.com")
hash.put("Sargeras", "sargeras@email.com")

print(hash.size())
print(hash.to_string())
